modules/core/src/main/java/org/apache/ignite/internal/processors/performancestatistics/FilePerformanceStatisticsWriter.java

```diff
@@ -0,0 +1,452 @@
+import java.util.function.IntSupplier;
+import org.apache.ignite.IgniteCheckedException;
+import org.apache.ignite.IgniteLogger;
+import org.apache.ignite.internal.GridKernalContext;
+import org.apache.ignite.internal.IgniteInterruptedCheckedException;
+import org.apache.ignite.internal.processors.cache.persistence.file.FileIO;
+import org.apache.ignite.internal.processors.cache.persistence.file.FileIOFactory;
+import org.apache.ignite.internal.processors.cache.persistence.file.RandomAccessFileIOFactory;
+import org.apache.ignite.internal.processors.cache.persistence.wal.SegmentedRingByteBuffer;
+import org.apache.ignite.internal.processors.cache.query.GridCacheQueryType;
+import org.apache.ignite.internal.util.GridIntIterator;
+import org.apache.ignite.internal.util.GridIntList;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.apache.ignite.internal.util.worker.GridWorker;
+import org.apache.ignite.lang.IgniteUuid;
+import org.apache.ignite.thread.IgniteThread;
+
+import static org.apache.ignite.internal.processors.performancestatistics.OperationType.JOB;
+import static org.apache.ignite.internal.processors.performancestatistics.OperationType.QUERY;
+import static org.apache.ignite.internal.processors.performancestatistics.OperationType.QUERY_READS;
+import static org.apache.ignite.internal.processors.performancestatistics.OperationType.TASK;
+import static org.apache.ignite.internal.processors.performancestatistics.OperationType.TX_COMMIT;
+import static org.apache.ignite.internal.processors.performancestatistics.OperationType.TX_ROLLBACK;
+
+/**
+ * Performance statistics writer based on logging to a file.
+ * <p>
+ * Each node collects statistics to a file placed under {@link #PERFORMANCE_STAT_DIR}.
+ * <p>
+ * <b>Note:</b> Start again will erase previous performance statistics files.
+ * <p>
+ * To iterate over records use {@link FilePerformanceStatisticsReader}.
+ */
+public class FilePerformanceStatisticsWriter {
+    /** Directory to store performance statistics files. Placed under Ignite work directory. */
+    public static final String PERFORMANCE_STAT_DIR = "performanceStatistics";
+
+    /** Default maximum file size in bytes. Performance statistics will be stopped when the size exceeded. */
+    public static final long DFLT_FILE_MAX_SIZE = 32 * U.GB;
+
+    /** Default off heap buffer size in bytes. */
+    public static final int DFLT_BUFFER_SIZE = (int)(32 * U.MB);
+
+    /** Default minimal batch size to flush in bytes. */
+    public static final int DFLT_FLUSH_SIZE = (int)(8 * U.MB);
+
+    /** Factory to provide I/O interface. */
+    private final FileIOFactory fileIoFactory = new RandomAccessFileIOFactory();
+
+    /** Performance statistics file writer worker. */
+    private final FileWriter fileWriter;
+
+    /** Performance statistics file I/O. */
+    private final FileIO fileIo;
+
+    /** File write buffer. */
+    private final SegmentedRingByteBuffer ringByteBuffer;
```

Comment:
we use `buffer` here but `buf` in `smallBufLogged`.
Let's use `buf` every where.

Reply:
Done.